server/license: TestRefreshLicenseEnforcerOnLicenseChange failed

server/license.TestRefreshLicenseEnforcerOnLicenseChange [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21206560?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21206560?buildTab=artifacts#/) on master @ [3b93a56edfd7bea885cfd9f84f3d874c2d549971](https://github.com/cockroachdb/cockroach/commits/3b93a56edfd7bea885cfd9f84f3d874c2d549971):


```
I260306 04:59:47.864397 19090 1@util/log/event_log.go:90  [T1,Vsystem,n1,client=127.0.0.1:46004,hostssl,user=root] 319 ={"Timestamp":1772773187864227744,"EventType":"set_cluster_setting","Statement":"SET CLUSTER SETTING \"cluster.organization\" = 'CRDB Unit Test'","Tag":"SET CLUSTER SETTING","User":"root","TxnReadTimestamp":1772773187862140533,"SettingName":"cluster.organization","Value":"'CRDB Unit Test'"}
I260306 04:59:58.053179 18920 sql/stats/automatic_stats.go:773  [T1,Vsystem,n1] 367  quiescing stats garbage collector
W260306 04:59:58.053242 18965 rpc/drpc.go:348  [-] 368  drpc server error context canceled
I260306 04:59:58.053334 19111 4@util/log/event_log.go:90  [T1,Vsystem,n1,client=127.0.0.1:46006,hostssl,user=root] 369 ={"Timestamp":1772773198053330523,"EventType":"client_session_end","InstanceID":1,"Network":"tcp","RemoteAddress":"127.0.0.1:46006","SessionID":"189a27d2626f6b830000000000000001","Duration":10181531576}
I260306 04:59:58.053377 14383 server/start_listen.go:114  [T1,Vsystem,n1] 370  server shutting down: instructing cmux to stop accepting
I260306 04:59:58.053395 19064 4@util/log/event_log.go:90  [T1,Vsystem,n1,client=127.0.0.1:46004,hostssl,user=root] 371 ={"Timestamp":1772773198053392250,"EventType":"client_session_end","InstanceID":1,"Network":"tcp","RemoteAddress":"127.0.0.1:46004","SessionID":"189a27d2618b598c0000000000000001","Duration":10196539633}
W260306 04:59:58.053433 17842 15@kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go:417  [T1,Vsystem,n1,rangefeed=lease,mux_n=1,gen=2] 372  error closing mux rangefeed client: context canceled
W260306 04:59:58.053451 18204 15@kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go:417  [T1,Vsystem,n1,rangefeed=statement-hints-watcher,mux_n=1,gen=2] 373  error closing mux rangefeed client: context canceled
W260306 04:59:58.053477 18712 15@kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go:417  [T1,Vsystem,n1,job=AUTO SPAN CONFIG RECONCILIATION id=1155736493810581505,rangefeed=sql-watcher-zones-rangefeed,mux_n=1,gen=2] 374  error closing mux rangefeed client: context canceled
W260306 04:59:58.053519 17719 sql/sqlliveness/slinstance/slinstance.go:337  [T1,Vsystem,n1] 375  exiting heartbeat loop
W260306 04:59:58.053528 17719 sql/sqlliveness/slinstance/slinstance.go:324  [T1,Vsystem,n1] 376  exiting heartbeat loop with error: node unavailable; try another peer
E260306 04:59:58.053548 17719 server/server_sql.go:535  [T1,Vsystem,n1] 377  failed to run update of instance with new session ID: node unavailable; try another peer
W260306 04:59:58.053662 17965 15@kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go:417  [T1,Vsystem,n1,rangefeed=settings-watcher,mux_n=1,gen=2] 378  error closing mux rangefeed client: context canceled
W260306 04:59:58.053774 16385 15@kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go:417  [T1,Vsystem,n1,rangefeed=tenant-settings-watcher,mux_n=1,gen=2] 379  error closing mux rangefeed client: context canceled
W260306 04:59:58.053849 18709 15@kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go:417  [T1,Vsystem,n1,job=AUTO SPAN CONFIG RECONCILIATION id=1155736493810581505,rangefeed=sql-watcher-descriptor-rangefeed,mux_n=1,gen=2] 380  error closing mux rangefeed client: context canceled
W260306 04:59:58.053879 19145 15@kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go:417  [T1,Vsystem,n1,rangefeed=tenant-entry-watcher,mux_n=1,gen=2] 381  error closing mux rangefeed client: context canceled
W260306 04:59:58.053919 18706 15@kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go:417  [T1,Vsystem,n1,job=AUTO SPAN CONFIG RECONCILIATION id=1155736493810581505,rangefeed=sql-watcher-protected-ts-records-rangefeed,mux_n=1,gen=2] 382  error closing mux rangefeed client: context canceled
W260306 04:59:58.053924 17852 15@kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go:417  [T1,Vsystem,n1,rangefeed=sql_instances,mux_n=1,gen=2] 383  error closing mux rangefeed client: context canceled
I260306 04:59:58.053960 18110 jobs/registry.go:1613  [T1,Vsystem,n1,job=1155736493810581505] 384  AUTO SPAN CONFIG RECONCILIATION job 1155736493810581505: stepping through state succeeded
W260306 04:59:58.054001 18946 15@kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go:417  [T1,Vsystem,n1,rangefeed=table-stats-cache,mux_n=1,gen=2] 386  error closing mux rangefeed client: context canceled
W260306 04:59:58.054057 19044 15@kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go:417  [T1,Vsystem,n1,rangefeed=tenant-boundaries-watcher,mux_n=1,gen=2] 387  error closing mux rangefeed client: context canceled
I260306 04:59:58.053992 18919 sql/stats/automatic_stats.go:721  [T1,Vsystem,n1] 385  quiescing auto stats refresher
W260306 04:59:58.054086 18391 15@kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go:417  [T1,Vsystem,n1,rangefeed=system-config-cache,mux_n=1,gen=2] 388  error closing mux rangefeed client: context canceled
W260306 04:59:58.054116 18063 15@kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go:417  [T1,Vsystem,n1,rangefeed=spanconfig-subscriber,mux_n=1,gen=2] 389  error closing mux rangefeed client: context canceled
W260306 04:59:58.059074 16501 server/server_sql.go:1841  [T1,Vsystem,n1] 390  server shutdown without a prior graceful drain
I260306 04:59:58.059212 16501 server/server_controller.go:416  [T1,Vsystem,n1] 391  server controller shutting down
I260306 04:59:58.059231 16501 server/server_controller.go:425  [T1,Vsystem,n1] 392  waiting for tenant servers to report stopped
--- FAIL: TestRefreshLicenseEnforcerOnLicenseChange (10.80s)
=== RUN   TestRefreshLicenseEnforcerOnLicenseChange/test_5
I260306 04:59:47.932670 19230 server/license/enforcer.go:553  [-] 349  trial license expiry was reset
I260306 04:59:47.935870 17963 server/license/enforcer.go:498  [T1,Vsystem,n1,rangefeed=settings-watcher] 350  enforcer license updated: type is none, grace period ends at "2024-09-21 12:28:36 +0000 UTC", telemetry required: false
I260306 04:59:47.935981 14792 1@util/log/event_log.go:90  [T1,Vsystem,n1,client=127.0.0.1:46006,hostssl,user=root] 351 ={"Timestamp":1772773187935757975,"EventType":"set_cluster_setting","Statement":"SET CLUSTER SETTING \"enterprise.license\" = ''","Tag":"SET CLUSTER SETTING","User":"root","TxnReadTimestamp":1772773187934083644,"SettingName":"enterprise.license"}
I260306 04:59:47.940727 17963 server/license/enforcer.go:498  [T1,Vsystem,n1,rangefeed=settings-watcher] 352  enforcer license updated: type is trial, grace period ends at "2000-01-08 12:00:00 +0000 UTC", expiration at "2000-01-01 12:00:00 +0000 UTC", telemetry required: true
I260306 04:59:47.941940 14792 1@util/log/event_log.go:90  [T1,Vsystem,n1,client=127.0.0.1:46006,hostssl,user=root] 353 ={"Timestamp":1772773187941872434,"EventType":"set_cluster_setting","Statement":"SET CLUSTER SETTING \"enterprise.license\" = 'crl-0-EMDYt8MDGAQiDkNSREIgVW5pdCBUZXN0'","Tag":"SET CLUSTER SETTING","User":"root","TxnReadTimestamp":1772773187939171643,"SettingName":"enterprise.license","Value":"crl-0-EMDYt8MDGAQiDkNSREIgVW5pdCBUZXN0"}
I260306 04:59:47.944659 17963 server/license/enforcer.go:536  [T1,Vsystem,n1,rangefeed=settings-watcher] 354  trial license expiry timestamp is 2000-01-01 12:00:00 +0000 UTC
W260306 04:59:47.947230 17963 server/settingswatcher/settings_watcher.go:475  [T1,Vsystem,n1,rangefeed=settings-watcher] 355  failed to set setting enterprise.license to crl-0-EPzYt8MDGAQiDkNSREIgVW5pdCBUZXN0: a trial license has previously been installed on this cluster
W260306 04:59:47.947230 17963 server/settingswatcher/settings_watcher.go:475  [T1,Vsystem,n1,rangefeed=settings-watcher] 355 +HINT: Please install a non-trial license to continue
I260306 04:59:48.369721 17157 1@gossip/gossip.go:1507  [T1,Vsystem,n1] 356  node has connected to cluster via gossip
I260306 04:59:48.369857 17157 13@kv/kvserver/stores.go:271  [T1,Vsystem,n1] 357  wrote 0 node addresses to persistent storage
I260306 04:59:53.439008 18110 13@spanconfig/spanconfigstore/store.go:364  [T1,Vsystem,n1,job=AUTO SPAN CONFIG RECONCILIATION id=1155736493810581505] 358  changing the spanconfig for span:/Table/4{2-3} from:{RangeMinBytes:134217728 RangeMaxBytes:536870912 GCPolicy:{TTLSeconds:14400 ProtectionPolicies:[] IgnoreStrictEnforcement:true} GlobalReads:false NumReplicas:1 NumVoters:0 Constraints:[] VoterConstraints:[] LeasePreferences:[] RangefeedEnabled:true ExcludeDataFromBackup:false} to:{RangeMinBytes:134217728 RangeMaxBytes:536870912 GCPolicy:{TTLSeconds:3600 ProtectionPolicies:[] IgnoreStrictEnforcement:true} GlobalReads:false NumReplicas:1 NumVoters:0 Constraints:[] VoterConstraints:[] LeasePreferences:[] RangefeedEnabled:true ExcludeDataFromBackup:false}
I260306 04:59:53.439130 18110 13@spanconfig/spanconfigstore/store.go:364  [T1,Vsystem,n1,job=AUTO SPAN CONFIG RECONCILIATION id=1155736493810581505] 359  changing the spanconfig for span:/Table/4{3-4} from:{RangeMinBytes:134217728 RangeMaxBytes:536870912 GCPolicy:{TTLSeconds:14400 ProtectionPolicies:[] IgnoreStrictEnforcement:true} GlobalReads:false NumReplicas:1 NumVoters:0 Constraints:[] VoterConstraints:[] LeasePreferences:[] RangefeedEnabled:true ExcludeDataFromBackup:false} to:{RangeMinBytes:134217728 RangeMaxBytes:536870912 GCPolicy:{TTLSeconds:3600 ProtectionPolicies:[] IgnoreStrictEnforcement:true} GlobalReads:false NumReplicas:1 NumVoters:0 Constraints:[] VoterConstraints:[] LeasePreferences:[] RangefeedEnabled:true ExcludeDataFromBackup:false}
I260306 04:59:53.439178 18110 13@spanconfig/spanconfigstore/store.go:364  [T1,Vsystem,n1,job=AUTO SPAN CONFIG RECONCILIATION id=1155736493810581505] 360  changing the spanconfig for span:/Table/5{8-9} from:{RangeMinBytes:134217728 RangeMaxBytes:536870912 GCPolicy:{TTLSeconds:14400 ProtectionPolicies:[] IgnoreStrictEnforcement:true} GlobalReads:false NumReplicas:1 NumVoters:0 Constraints:[] VoterConstraints:[] LeasePreferences:[] RangefeedEnabled:true ExcludeDataFromBackup:false} to:{RangeMinBytes:134217728 RangeMaxBytes:536870912 GCPolicy:{TTLSeconds:3600 ProtectionPolicies:[] IgnoreStrictEnforcement:true} GlobalReads:false NumReplicas:1 NumVoters:0 Constraints:[] VoterConstraints:[] LeasePreferences:[] RangefeedEnabled:true ExcludeDataFromBackup:false}
I260306 04:59:53.439217 18110 13@spanconfig/spanconfigstore/store.go:364  [T1,Vsystem,n1,job=AUTO SPAN CONFIG RECONCILIATION id=1155736493810581505] 361  changing the spanconfig for span:/Table/6{1-2} from:{RangeMinBytes:134217728 RangeMaxBytes:536870912 GCPolicy:{TTLSeconds:14400 ProtectionPolicies:[] IgnoreStrictEnforcement:true} GlobalReads:false NumReplicas:1 NumVoters:0 Constraints:[] VoterConstraints:[] LeasePreferences:[] RangefeedEnabled:true ExcludeDataFromBackup:false} to:{RangeMinBytes:134217728 RangeMaxBytes:536870912 GCPolicy:{TTLSeconds:3600 ProtectionPolicies:[] IgnoreStrictEnforcement:true} GlobalReads:false NumReplicas:1 NumVoters:0 Constraints:[] VoterConstraints:[] LeasePreferences:[] RangefeedEnabled:true ExcludeDataFromBackup:false}
I260306 04:59:53.439256 18110 13@spanconfig/spanconfigstore/store.go:364  [T1,Vsystem,n1,job=AUTO SPAN CONFIG RECONCILIATION id=1155736493810581505] 362  changing the spanconfig for span:/Table/6{2-3} from:{RangeMinBytes:134217728 RangeMaxBytes:536870912 GCPolicy:{TTLSeconds:14400 ProtectionPolicies:[] IgnoreStrictEnforcement:true} GlobalReads:false NumReplicas:1 NumVoters:0 Constraints:[] VoterConstraints:[] LeasePreferences:[] RangefeedEnabled:true ExcludeDataFromBackup:false} to:{RangeMinBytes:134217728 RangeMaxBytes:536870912 GCPolicy:{TTLSeconds:3600 ProtectionPolicies:[] IgnoreStrictEnforcement:true} GlobalReads:false NumReplicas:1 NumVoters:0 Constraints:[] VoterConstraints:[] LeasePreferences:[] RangefeedEnabled:true ExcludeDataFromBackup:false}
I260306 04:59:57.391259 11712 2@util/log/event_log.go:90  [T1,Vsystem,n1] 363 ={"Timestamp":1772773197391254720,"EventType":"runtime_stats","MemRSSBytes":491352064,"GoroutineCount":672,"MemStackSysBytes":20545536,"GoAllocBytes":223244552,"GoTotalBytes":458160488,"HeapFragmentBytes":58224376,"HeapReservedBytes":137592832,"HeapReleasedBytes":9183232,"CGoAllocBytes":16693024,"CGoTotalBytes":66715648,"CGoCallRate":0.00000039655382,"GCPausePercent":0.0000000027173264,"GCRunCount":13,"NetHostRecvBytes":134028084,"NetHostSendBytes":134028084}
I260306 04:59:57.391406 11712 2@server/status/runtime_log.go:43  [T1,Vsystem,n1] 364  runtime stats: 469 MiB RSS, 672 goroutines (stacks: 20 MiB), 213 MiB/437 MiB Go alloc/total (heap fragmentation: 56 MiB, heap reserved: 131 MiB, heap released: 8.8 MiB), 16 MiB/64 MiB CGO alloc/total (0.0 CGO/sec), 0.0/0.0 %(u/s)time, 0.0 %gc (13x), 128 MiB/128 MiB (r/w)net
I260306 04:59:57.407847 17332 3@pebble/event.go:1308  [n1,s1,pebble] 365  [JOB 6] WAL created 000007
W260306 04:59:58.052601 14792 sql/set_cluster_setting.go:718  [T1,Vsystem,n1,client=127.0.0.1:46006,hostssl,user=root] 366  SET CLUSTER SETTING "enterprise.license" timed out waiting for value "crl-0-EPzYt8MDGAQiDkNSREIgVW5pdCBUZXN0", observed "crl-0-EMDYt8MDGAQiDkNSREIgVW5pdCBUZXN0"
    license_enforcer_integration_test.go:142: expected query 'SET CLUSTER SETTING enterprise.license = 'crl-0-EPzYt8MDGAQiDkNSREIgVW5pdCBUZXN0'' error 'a trial license has previously been installed on this cluster', got: pq: setting updated but timed out waiting to read new value
--- FAIL: TestRefreshLicenseEnforcerOnLicenseChange/test_5 (10.12s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRefreshLicenseEnforcerOnLicenseChange.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-61101

--- a/pkg/server/license/enforcer.go
+++ b/pkg/server/license/enforcer.go
@@ -82,11 +82,6 @@ type Enforcer struct {
 	// are bypassed. This is typically used to disable enforcement for single-node deployments.
 	isDisabled atomic.Bool
 
-	// trialUsageExpiry records the expiration timestamp, in seconds, of any
-	// trial license on this cluster (past or present). A value of 0 indicates
-	// that no trial license has ever been installed.
-	trialUsageExpiry atomic.Int64
-
 	// db is a pointer to the database for use for KV read/writes. This is only
 	// set for the system tenant.
 	db isql.DB
@@ -255,12 +250,12 @@ func (e *Enforcer) initClusterMetadata(ctx context.Context, options options) err
 			return err
 		}
 		if trialUsageCount.Value == nil {
-			e.trialUsageExpiry.Store(0)
+			trialLicenseExpiryTimestamp.Store(0)
 		} else {
-			e.trialUsageExpiry.Store(trialUsageCount.ValueInt())
+			trialLicenseExpiryTimestamp.Store(trialUsageCount.ValueInt())
 		}
 		log.Dev.Infof(ctx, "trial license expiry initialized to %s",
-			timeutil.Unix(e.trialUsageExpiry.Load(), 0))
+			timeutil.Unix(trialLicenseExpiryTimestamp.Load(), 0))
 
 		// Cache and maybe set the cluster init grace period timestamp. This is the
 		// grace period we will use if the cluster does not have a license installed.
@@ -498,20 +493,20 @@ func (e *Enforcer) RefreshForLicenseChange(
 	log.Dev.Infof(ctx, "%s", sb.RedactableString())
 }
 
-// UpdateTrialLicenseExpiry returns the expiration timestamp of any trial license
-// used on the cluster, including the new trial license if a change is being applied.
-// This function updates the expiry if the current license is changing and is a trial.
+// UpdateTrialLicenseExpiry updates the expiration timestamp of trial license
+// usage on the cluster. This function updates the expiry if the current license
+// is changing and is a trial.
 func (e *Enforcer) UpdateTrialLicenseExpiry(
 	ctx context.Context, currentLicense LicType, isLicenseChange bool, expiry int64,
-) (curExpiry int64, err error) {
-	// If we aren't setting a new trial license, return the cached copy. The e.db
-	// check is necessary as that's needed to read/write to the KV. This will be
-	// set for the system tenant, which is where the license can ever be set anyway.
+) error {
+	// If we aren't setting a new trial license, return early. The e.db check is
+	// necessary as that's needed to read/write to the KV. This will be set for
+	// the system tenant, which is where the license can ever be set anyway.
 	if currentLicense != LicTypeTrial || !isLicenseChange || e.db == nil {
-		return e.trialUsageExpiry.Load(), nil
+		return nil
 	}
 
-	err = e.db.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+	err := e.db.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
 		// We only allow a single trial license to be installed. If one is
 		// already set in the KV, exit and return that expiry value.
 		oldVal, err := txn.KV().Get(ctx, keys.TrialLicenseExpiry)
@@ -519,22 +514,23 @@ func (e *Enforcer) UpdateTrialLicenseExpiry(
 			return err
 		}
 		if oldVal.Value != nil && oldVal.ValueInt() > 0 {
-			e.trialUsageExpiry.Store(oldVal.ValueInt())
+			trialLicenseExpiryTimestamp.Store(oldVal.ValueInt())
 			return nil
 		}
 		err = txn.KV().Put(ctx, keys.TrialLicenseExpiry, expiry)
 		if err != nil {
 			return err
 		}
-		e.trialUsageExpiry.Store(expiry)
+		trialLicenseExpiryTimestamp.Store(expiry)
 		return nil
 	})
 	if err != nil {
-		return
+		return err
 	}
-	curExpiry = e.trialUsageExpiry.Load()
-	log.Dev.Infof(ctx, "trial license expiry timestamp is %s", timeutil.Unix(curExpiry, 0))
-	return
+	curExpiry := trialLicenseExpiryTimestamp.Load()
+	log.Dev.Infof(ctx, "trial license expiry timestamp is %s",
+		timeutil.Unix(curExpiry, 0))
+	return nil
 }
 
 // TestingResetTrialUsage is an API to clear the license expiry in the KV. This is only used
@@ -548,7 +544,6 @@ func (e *Enforcer) TestingResetTrialUsage(ctx context.Context) error {
 		if err != nil {
 			return err
 		}
-		e.trialUsageExpiry.Store(0)
 		trialLicenseExpiryTimestamp.Store(0)
 		log.Dev.Infof(ctx, "trial license expiry was reset")
 		return nil
@@ -558,7 +553,7 @@ func (e *Enforcer) TestingResetTrialUsage(ctx context.Context) error {
 // TestingGetTrialUsageExpiry returns the trial usage expiry timestamp for
 // testing purposes.
 func (e *Enforcer) TestingGetTrialUsageExpiry() int64 {
-	return e.trialUsageExpiry.Load()
+	return trialLicenseExpiryTimestamp.Load()
 }
 
 // TestingMaybeFailIfThrottled is a helper that runs MaybeFailIfThrottled in a separate goroutine.

--- a/pkg/server/license/license.go
+++ b/pkg/server/license/license.go
@@ -194,14 +194,13 @@ func registerCallbackOnLicenseChange(
 		}
 		licenseEnforcer.RefreshForLicenseChange(ctx, licenseType, licenseExpiry)
 
-		expiry, err := licenseEnforcer.UpdateTrialLicenseExpiry(
+		err = licenseEnforcer.UpdateTrialLicenseExpiry(
 			ctx, licenseType, isChange, licenseExpiry.Unix())
 		if err != nil {
 			log.Dev.Errorf(ctx,
 				"unable to update trial license expiry: %v", err)
 			return
 		}
-		trialLicenseExpiryTimestamp.Store(expiry)
 	}
 	// Install the hook so that we refresh license details when the license
 	// changes.

